### PR TITLE
[AppConfig] BREAKING CHANGE: Update default connection string resolution logic

### PR DESCRIPTION
**Pull request by albertofori**
**Related command**
`az appconfig`

**Description**

This update ensures that command line arguments take precedence over defaults configured using `az configure` or in environment variables to avoid conflicts possible conflict errors.

**Testing Guide**

**History Notes**

[AppConfig] BREAKING CHANGE: `az appconfig`: Update default connection string resolution logic

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [x] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

- [x] I adhere to the [Error Handling Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/error_handling_guidelines.md).

### CODE REVIEW

**Comment by yonzhan** [comment]
AppConfig

**Comment by albertofori** [comment]
Hi @avanigupta , could you please take a look at this?

**Comment by avanigupta** [comment]
@albertofori, we have similar defaults for store name, endpoint and auth-mode (there might be others I'm missing). Can you help to add the same behavior across all our default arguments?

**Comment by albertofori** [comment]
> @albertofori, we have similar defaults for store name, endpoint and auth-mode (there might be others I'm missing). Can you help to add the same behavior across all our default arguments?

Hi @avanigupta, as you mentioned, the other arguments have defaults and the resolution of these defaults are already handled correctly in code. It appears only the connection string has the current issue since it is the only value that we attempt to get from environment variables or azure default configuration when not supplied.

**Comment by albertofori** [comment]
> @albertofori, we have similar defaults for store name, endpoint and auth-mode (there might be others I'm missing). Can you help to add the same behavior across all our default arguments?

Hi @avanigupta, it appears the values that have defaults are automatically handled in the preferred way, however, this issue arose with connection strings due to the dependency between the `connection-string` and `name`

File: src/azure-cli/azure/cli/command_modules/appconfig/_params.py [hunk 1]
```diff
@@ -79,10 +79,19 @@ def load_arguments(self, _):
         configured_default=None
     )
 
+    # Used with data plane commands. These take either a store name or connection string argument.
+    # We only read default values when neither connection string nor store name is provided so configured defaults are not supplied.
+    data_plane_name_arg_type = CLIArgumentType(
+        options_list=['--name', '-n'],
+        type=str,
+        help='Name of the App Configuration. You can configure the default name using `az configure --defaults app_configuration_store=<name>`.',
+        configured_default=None
```
[the remark this answers is not in the document]
Comment (albertofori) [reply src/azure-cli/azure/cli/command_modules/appconfig/_params.py:88]: Without explicitly specifying the configured default as `None` it is assigned the app_configuration_store default value as specified in the [default](https://github.com/Azure/azure-cli/blob/80f487d1241deff4c7d314e18cb54bef16a695bb/src/azure-cli/azure/cli/command_modules/appconfig/_params.py#:~:text=%27name%27%2C%20options_list) `name` argument.

File: src/azure-cli/azure/cli/command_modules/appconfig/_validators.py [hunk 2]
```diff
@@ -30,8 +30,13 @@ def validate_datetime(namespace):
             'The input datetime is invalid. Correct format should be YYYY-MM-DDThh:mm:ssZ ')
 
 
-def validate_connection_string(namespace):
+def validate_connection_string(cmd, namespace):
     ''' Endpoint=https://example.azconfig.io;Id=xxxxx;Secret=xxxx'''
+    # Only use defaults when both name and connection string not specified
+    if not (namespace.connection_string or namespace.name):
+        namespace.connection_string = cmd.cli_ctx.config.get('appconfig', 'connection_string', None) or cmd.cli_ctx.config.get('defaults', 'appconfig_connection_string', None)
```
[the remark this answers is not in the document]
Comment (albertofori) [reply src/azure-cli/azure/cli/command_modules/appconfig/_validators.py]: This has been updated. Thanks!

**Review by avanigupta — DISMISSED** [review]

**Comment by avanigupta** [comment]
This might be a breaking change for some users so please make sure it appears in the release notes.

**Comment by zhoxing-ms** [comment]
> [AppConfig] BREAKING CHANGE: az appconfig: Update default connection string resolution logic

@albertofori In order to reduce the impact of breaking change on users, we usually recommend that breaking change be placed in the breaking change window *(twice a year, the date of the next window is in Build Sprint, about May)*
For more information about the process of breaking change, please refer to document [breaking_changes_doc](https://eng.ms/docs/cloud-ai-platform/azure-core/azure-management-and-platforms/control-plane-bburns/azure-cli-tools-azure-cli-powershell-and-terraform/azure-cli-tools/teams_docs/azcli_docs/breaking_changes)

So I'd like to confirm with you whether we can avoid this breaking change before the next breaking change window?

**Comment by albertofori** [comment]
> > [AppConfig] BREAKING CHANGE: az appconfig: Update default connection string resolution logic
> 
> @albertofori In order to reduce the impact of breaking change on users, we usually recommend that breaking change be placed in the breaking change window _(twice a year, the date of the next window is in Build Sprint, about May)_ For more information about the process of breaking change, please refer to document [breaking_changes_doc](https://eng.ms/docs/cloud-ai-platform/azure-core/azure-management-and-platforms/control-plane-bburns/azure-cli-tools-azure-cli-powershell-and-terraform/azure-cli-tools/teams_docs/azcli_docs/breaking_changes)
> 
> So I'd like to confirm with you whether we can avoid this breaking change before the next breaking change window?

@zhoxing-ms Yes please. I believe we could add this in the next breaking change window. Thanks!

**Comment by albertofori** [comment]
> > > [AppConfig] BREAKING CHANGE: az appconfig: Update default connection string resolution logic
> > 
> > 
> > @albertofori In order to reduce the impact of breaking change on users, we usually recommend that breaking change be placed in the breaking change window _(twice a year, the date of the next window is in Build Sprint, about May)_ For more information about the process of breaking change, please refer to document [breaking_changes_doc](https://eng.ms/docs/cloud-ai-platform/azure-core/azure-management-and-platforms/control-plane-bburns/azure-cli-tools-azure-cli-powershell-and-terraform/azure-cli-tools/teams_docs/azcli_docs/breaking_changes)
> > So I'd like to confirm with you whether we can avoid this breaking change before the next breaking change window?
> 
> @zhoxing-ms Yes please. I believe we could add this in the next breaking change window. Thanks!

@zhoxing-ms  Can this change go out as as discussed earlier?

**Comment by zhoxing-ms** [comment]
@albertofori Could you please resolve these conflicts?

**Comment by albertofori** [comment]
> @albertofori Could you please resolve these conflicts?

Hi @zhoxing-ms Please a change has been pushed to resolve the merge conflicts, thanks!

**Comment by zhoxing-ms** [comment]
@albertofori Thanks for the quick response, we will release this PR in the next breaking change release *(on 05-23)*

**Comment by zhoxing-ms** [comment]
@albertofori Albert Ofori FTE Could you please resolve these conflicts?

**Comment by albertofori** [comment]
> @albertofori Albert Ofori FTE Could you please resolve these conflicts?

@zhoxing-ms Please the conflicts have been resolved. Thanks!

**Review by zhoxing-ms — APPROVED** [review]

**Pull request merged**